### PR TITLE
Port from GObject to GLib and Python 2 to 3

### LINKED ISSUE #9: Port from GObject to GLib

### LINKED ISSUE #10: PyGI warnings
````
PyGIWarning: Gtk was imported without specifying a version first. Use gi.require_version('Gtk', '3.0') before import to ensure that the right version gets loaded.
  from gi.repository import Gtk, GObject, Gdk
/home/ayush/activities/locosugar/sprites.py:81: PyGIWarning: PangoCairo was imported without specifying a version first. 
Use gi.require_version('PangoCairo', '1.0') before import to ensure that the right version gets loaded.
  from gi.repository import Pango, PangoCairo
````

### LINKED ISSUE #15: Port to Python 3.

### PR DESCRIPTION
**Pull request by Saumya-Mishra9129**
Fixes #9, Fixes #15, Fixes #10 
@chimosky  Kindly review changes.

### CODE REVIEW

File: game.py [hunk 1]
```diff
@@ -398,7 +399,7 @@ def _move_loco(self, x, y, i):
     def _keypress_cb(self, area, event):
         ''' Keypress '''
         # Games 4, 5, and 6 use the keyboard
-        print 'keypress event'
+        print('keypress event')
```
Comment (chimosky) [inline game.py]: You can remove this line entirely as it's not needed.

File: game.py [hunk 2]
```diff
@@ -89,7 +90,7 @@ def __init__(self, canvas, parent=None, path=None):
 
         self._width = Gdk.Screen.width()
         self._height = Gdk.Screen.height()
-        self._scale = self._width / 1200.
+        self._scale = self._width // 1200.
```
Comment (chimosky) [inline game.py]: As one of the values here is a float, use `/` and not `//` so as to use the same values that were used before the port.

File: game.py [hunk 3]
```diff
@@ -221,13 +222,13 @@ def _all_clear(self):
     def _show_time(self):
         self.level = 0
         self._all_clear()
-        x = int(self._width / 4.)
-        y = int(self._height / 8.)
+        x = int(self._width // 4.)
+        y = int(self._height // 8.)
```
Comment (chimosky) [inline game.py]: Same here.

File: game.py [hunk 4]
```diff
@@ -221,13 +222,13 @@ def _all_clear(self):
...
         for i in range(len(str(self.score))):
             self._sticky_cards[i].move((x, y))
             self._sticky_cards[i].set_layer(LOCO_LAYER)
             self._sticky_cards[i].set_label(str(self.score)[i])
-            x += int(self._loco_dim[0] / 2.)
+            x += int(self._loco_dim[0] // 2.)
```
Comment (chimosky) [inline game.py]: And here.

File: game.py [hunk 5]
```diff
@@ -331,7 +332,7 @@ def new_game(self, first_time):
                     self._sticky_cards[c].set_layer(LOCO_LAYER)
                     self._sticky_cards[c].set_label(MSGS[self._counter][i])
                     c += 1
-                    x += int(self._loco_dim[0] / 2.)
+                    x += int(self._loco_dim[0] // 2.)
```
Comment (chimosky) [inline game.py]: And here.

File: game.py [hunk 6]
```diff
@@ -343,12 +344,12 @@ def new_game(self, first_time):
                 self._taunt(x, y, 0)
 
     def _quad_to_xy(self, q):
-        x = int(max(0, (self._width / 2.) * uniform(0, 1) - self._loco_dim[0]))
+        x = int(max(0, (self._width // 2.) * uniform(0, 1) - self._loco_dim[0]))
```
Comment (chimosky) [inline game.py]: Here too.

File: game.py [hunk 7]
```diff
@@ -343,12 +344,12 @@ def new_game(self, first_time):
...
         if q in [0, 1]:
-            x += int(self._width / 2.)
-        y = int(max(0, (self._height / 2.) * uniform(0, 1) - self._loco_dim[1]))
+            x += int(self._width // 2.)
+        y = int(max(0, (self._height // 2.) * uniform(0, 1) - self._loco_dim[1]))
```
Comment (chimosky) [inline game.py]: And here.

File: game.py [hunk 8]
```diff
@@ -343,12 +344,12 @@ def new_game(self, first_time):
...
         if q in [1, 2]:
-            y += int(self._height / 2.)
+            y += int(self._height // 2.)
```
Comment (chimosky) [inline game.py]: And here.

File: game.py [hunk 9]
```diff
@@ -494,21 +495,21 @@ def _keypress_cb(self, area, event):
...
-        x, y = map(int, event.get_coords())
+        x, y = list(map(int, event.get_coords()))
         if self._seconds > 1:
             self._panel.hide()
         if not self._clicked and self.level == 0:
             # For Game 0, see if the mouse is on the Loco
-            dx = x - self._loco_pos[0] - self._loco_dim[0] / 2.
-            dy = y - self._loco_pos[1] - self._loco_dim[1] / 2.
+            dx = x - self._loco_pos[0] - self._loco_dim[0] // 2.
+            dy = y - self._loco_pos[1] - self._loco_dim[1] // 2.
```
Comment (chimosky) [inline game.py]: Here too.

File: game.py [hunk 10]
```diff
@@ -494,21 +495,21 @@ def _keypress_cb(self, area, event):
...
             if dx * dx + dy * dy < 200:
                 self._clicked = True
                 if self._timeout_id is not None:
                     GLib.source_remove(self._timeout_id)
                 # Play again
                 self._all_clear()
-                self._man_cards[0].move((x - int(self._loco_dim[0] / 2.),
-                                         y - int(self._loco_dim[1] / 2.)))
+                self._man_cards[0].move((x - int(self._loco_dim[0] // 2.),
+                                         y - int(self._loco_dim[1] // 2.)))
```
Comment (chimosky) [inline game.py]: And here.

File: game.py [hunk 11]
```diff
@@ -531,7 +532,7 @@ def _mouse_move_cb(self, win, event):
             dy = y - self._drag_pos[1]
             self._press.move_relative((dx, dy))
             self._drag_pos = [x, y]
-            if x > self._width / 2.:
+            if x > self._width // 2.:
```
Comment (chimosky) [inline game.py]: And here.

File: game.py [hunk 12]
```diff
@@ -568,8 +569,8 @@ def _button_press_cb(self, win, event):
         # Games 1, 2, and 3 involve clicks; Games 4 and 5 allow click to drag
         if self.level == 1:
             self._all_clear()
-            self._man_cards[0].move((x - int(self._loco_dim[0] / 2.),
-                                     y - int(self._loco_dim[1] / 2.)))
+            self._man_cards[0].move((x - int(self._loco_dim[0] // 2.),
+                                     y - int(self._loco_dim[1] // 2.)))
```
Comment (chimosky) [inline game.py]: And here.

**Review by chimosky — CHANGES_REQUESTED** [review]

**Comment by chimosky** [comment]
Nice work, I think you can improve on your commit messages, [making commits](https://github.com/sugarlabs/sugar-docs/blob/master/src/contributing.md#making-commits) should help.

Edited your opening comment, adding `Fixes #issueno` automatically closes the issue once the PR is merged and that makes the job easier.

**Comment by quozl** [comment]
Thanks.  Tested.

* [x] during press of lowercase keys, two locos occupied the same coordinates, so it was difficult to read the letters,
* [x] sometimes a letter for a loco may not be visible against the background; especially the fur of the dog,
* [x] the toolbar was invisible until the first successful set of rounds,
* [x] in one of the rounds, a loco became a ghost (green and purple) and stayed on screen for every subsequent round,
* [x] the toolbar shifted the display down enough such that one of the locos with text could not be fully displayed, so I didn't know what key to press to dismiss it,

Having now experienced it, the toolbar adds no value that couldn't be added by a stop or quit button instead.

Activity logs were noisy, frequent mentions of "keypress event" and 'WARNING: erroneous pipeline: syntax error".

**Comment by Saumya-Mishra9129** [comment]
@quozl  "keypress event" log's noise has been fixed. Working on testing .
@chimosky changes requested by you have been fixed.

**Comment by quozl** [comment]
Thanks.

**Comment by Saumya-Mishra9129** [comment]
> sometimes a letter for a loco may not be visible against the background; especially the fur of the dog,

Can we change the colour of letter for a loco from white to something else in order to avoid this.?

**Comment by Saumya-Mishra9129** [comment]
@quozl  

> the toolbar was invisible until the first successful set of rounds,

This is fixed in 3d64076.

> in one of the rounds, a loco became a ghost (green and purple) and stayed on screen for every subsequent round,

I am not getting this one on testing.

**Comment by quozl** [comment]
Thanks for the update.  Is 3d64076 correct way to solve this?  Is the activity better always fullscreen, and never show the toolbar?  We have quite a few activities that do this.  My observation in [earlier testing](https://github.com/sugarlabs/locosugar/pull/14#issuecomment-603599115) is that the toolbar did appear without being asked for, and also that it affected play.  Would a better way to solve this be to find out what is causing fullscreen mode to turn off, and stop doing that?

**Comment by Saumya-Mishra9129** [comment]
> Thanks for the update. Is [3d64076](https://github.com/sugarlabs/locosugar/commit/3d640769c704ea946e236c09008c9244b5bb5b9a) correct way to solve this? Is the activity better always fullscreen, and never show the toolbar? We have quite a few activities that do this. My observation in [earlier testing](https://github.com/sugarlabs/locosugar/pull/14#issuecomment-603599115) is that the toolbar did appear without being asked for, and also that it affected play. Would a better way to solve this be to find out what is causing fullscreen mode to turn off, and stop doing that?

Thanks @quozl for review. Actually I thought you are saying the reverse of it. I thought we have to display toolbar every time. So I have done that. There was confusion. 
According to me , It is better to have full screen always and show the toolbar when we want a toolbar by clicking. 

> the toolbar shifted the display down enough such that one of the locos with text could not be fully displayed, so I didn't know what key to press to dismiss it,

>  the toolbar was invisible until the first successful set of rounds,

aa9fd66 should fixed these both.

**Comment by Saumya-Mishra9129** [comment]
@quozl Also consider  [this](https://github.com/sugarlabs/locosugar/pull/14#issuecomment-608506787), What we need to do in this case. According to me changing color is best way for solving this.

**Comment by chimosky** [comment]
> According to me , It is better to have full screen always and show the toolbar when we want a toolbar by clicking.

How do I test? Clicked and toolbar still isn't visible.

**Comment by Saumya-Mishra9129** [comment]
> > According to me , It is better to have full screen always and show the toolbar when we want a toolbar by clicking.
> 
> How do I test? Clicked and toolbar still isn't visible.

@chimosky This was the main goal to make toolbar not visible because it was affecting the play.

**Comment by quozl** [comment]
> @quozl Also consider [this](https://github.com/sugarlabs/locosugar/pull/14#issuecomment-608506787), What we need to do in this case. According to me changing color is best way for solving this.

I'm not sure changing colour alone is best way.

Number or letter is to be distinguished from the background image.  There are several ways to do this.  Here are some of them;

* draw the number or letter using an XOR or inversion rather than in a specific colour,
* choose a colour for the number or letter based on a calculation of the average brightness in the vicinity of the drawing,
* reserve one range of brightnesses for background image (e.g. 0% to 80%), and another range of brightnesses for the number or letter (e.g.100% only).

I'm not going to be prescriptive on which way to use.  You may think up some other method.

**Comment by quozl** [comment]
> @chimosky This was the main goal to make toolbar not visible because it was affecting the play.

I'm fine with a toolbar, or without a toolbar, as long as these conditions can be satisfied;
* the toolbar is either present or absent, and doesn't change between states without explicit command from the user,
* the loco is not positioned in a way that hides the number or letter, (this happened in my earlier test once the toolbar became visible, in turn because the canvas was shifted downward and the positioning did not take that into account),
* the behaviour of the toolbar is not changed without some reason, otherwise teachers or children who upgrade the activity face a change they did not expect.

**Comment by Saumya-Mishra9129** [comment]
@quozl  

> during press of lowercase keys, two locos occupied the same coordinates, so it was difficult to read the letters,

I am not sure how to fix this issue but as I looked at code of game.py, it uses random.uniform for generating location of a loco, as numbers are generated randomly and uniform method having very less range, it can possibly increases chances of collapsing of two locos. and Also I found same issue during press of uppercase keys. It may be possible that increasing the range of uniform can solve this, but I am not sure about it.

**Comment by quozl** [comment]
I've not examined the code.  A collision detection algorithm should be used; so that when a location is randomly generated it is checked to see if it is close to another location already used, and if so the algorithm would try again.

**Comment by quozl** [comment]
I've examined the code, the history of commits, and made some more tests.  The collision is expected and the solution is to drag the Loco.  The same for when the letter under the Loco is drawn against a white background.  I've marked those test results as resolved.

**Comment by quozl** [comment]
Tested e013fd5.

* [ ] the "WARNING: erroneous pipeline: syntax error" continues to occur,
* [x] the toolbar appears at the end of the game and can't be dismissed; if the child continues to play the game there's possibility of Locos positioned such that their letters are invisible, but they can be dragged.  Should the toolbar appear?

**Comment by Saumya-Mishra9129** [comment]
@quozl You were supposed to test the latest commit aa9fd66 .
When I checked with aa9fd66 , 

> the toolbar appears at the end of the game and can't be dismissed; if the child continues to play the game there's possibility of Locos positioned such that their letters are invisible, but they can be dragged. Should the toolbar appear?

The toolbar does not appear at the end of the game. I suggest you to check once with  aa9fd66 .
Also I am not getting the WARNING: erroneous pipeline: syntax error anywhere. I am not able to understand why it is happening.

**Comment by quozl** [comment]
Thanks.  Tested aa9fd66 on Ubuntu 20.04 on an OLPC NL3 laptop.  The toolbar did not appear.

* [ ] the background does not cover the display, which is 1366x768 on this laptop, and the uncovered area is grey.  The greeting message window is not fully within the display (because the vertical size of the display is much less than the 900 pixels the program is designed for).  The moving Loco in the first level of the game limits position to the area of the background.  The drag to the right game used the centre of display rather than the centre of the background image.

Screenshot of _LocoSugar Activity_

* [ ] "WARNING: erroneous pipeline: syntax error" does occur, and by using `strace -ff -o ~/ls -s 64 -i sugar-activity3`, followed by a search for the message in the strace logs, followed by analysis of the logs, found that the processes that emit the message were running the program `/usr/bin/gst-launch-1.0`, which is called by `play_audio.py` making a subprocess call with an incorrect pipeline; there's also no sound heard.  I recommend the player module be replaced with a more modern one that calls the PyGObject Gst module instead; you can find several in maintained activities throughout Sugar Labs.

```log
$ grep WARNING ls*
ls.1794:[00007f4c7e6e9057] write(2, "WARNING: erroneous pipeline: syntax error\n",
42) = 42
$ grep execve ls.1794 | grep -v ENOENT
ls.1794:[00007f10b3df216b] execve("/usr/bin/gst-launch-1.0", ["gst-launch-1.0", "filesrc",
"location=/usr/share/sugar/activities/LocoSugar.activity/sounds/d"..., "! oggdemux",
"! vorbisdec", "! audioconvert", "! alsasink"], 0x11cfb30 /* 64 vars */) = 0
```

* [ ] the ghost Loco happened again,

**Comment by Saumya-Mishra9129** [comment]
> the background does not cover the display, which is 1366x768 on this laptop, and the uncovered area is grey. The greeting message window is not fully within the display (because the vertical size of the display is much less than the 900 pixels the program is designed for). The moving Loco in the first level of the game limits position to the area of the background. The drag to the right game used the centre of display rather than the centre of the background image.

I am not getting as we need to fix only for 1200x900 resolution or for all other systems as well. ` the background does not cover the display, which is 1366x768 on this laptop, and the uncovered area is grey. ` this could happen because the width is greater than 1200 for what it is designed for.

**Comment by quozl** [comment]
I can see how you could get confused.  Many Sugar activities were designed for a fixed resolution of 1200x900 on the OLPC XO laptop, with no intention at the time for the activity to be resolution independent, adaptable, or responsive.  This is certainly one of those kinds of activities, as it has bitmap instead of vector graphics.

The OLPC XO laptop is unlikely to run Python 3 activities because of the huge cost of upgrading the rest of the software stack; firmware, kernel, graphics drivers, and so forth.  So as soon as an activity tries to transition from Python 2 to Python 3, it brings with it (a) not having to maintain the 1200x900 resolution support, and (b) having to maintain other resolutions and responsiveness.

There are many ways to adapt this activity to other resolutions; best solution is one that works on all expected resolutions, by calculating where to place the pixmaps, and how to either wrap, tile, or place them within letterboxes.

**Comment by Saumya-Mishra9129** [comment]
> Thanks. Tested [aa9fd66](https://github.com/sugarlabs/locosugar/commit/aa9fd66d9276deb51ff85823dfe0d777dffbe5da) on Ubuntu 20.04 on an OLPC NL3 laptop. The toolbar did not appear.
> 
> * [ ]  the background does not cover the display, which is 1366x768 on this laptop, and the uncovered area is grey.  The greeting message window is not fully within the display (because the vertical size of the display is much less than the 900 pixels the program is designed for).  The moving Loco in the first level of the game limits position to the area of the background.  The drag to the right game used the centre of display rather than the centre of the background image.
> 
> Screenshot of LocoSugar Activity
> 
> * [ ]  "WARNING: erroneous pipeline: syntax error" does occur, and by using `strace -ff -o ~/ls -s 64 -i sugar-activity3`, followed by a search for the message in the strace logs, followed by analysis of the logs, found that the processes that emit the message were running the program `/usr/bin/gst-launch-1.0`, which is called by `play_audio.py` making a subprocess call with an incorrect pipeline; there's also no sound heard.  I recommend the player module be replaced with a more modern one that calls the PyGObject Gst module instead; you can find several in maintained activities throughout Sugar Labs.
> 
> ```
> $ grep WARNING ls*
> ls.1794:[00007f4c7e6e9057] write(2, "WARNING: erroneous pipeline: syntax error\n",
> 42) = 42
> $ grep execve ls.1794 | grep -v ENOENT
> ls.1794:[00007f10b3df216b] execve("/usr/bin/gst-launch-1.0", ["gst-launch-1.0", "filesrc",
> "location=/usr/share/sugar/activities/LocoSugar.activity/sounds/d"..., "! oggdemux",
> "! vorbisdec", "! audioconvert", "! alsasink"], 0x11cfb30 /* 64 vars */) = 0
> ```
> 
> * [ ]  the ghost Loco happened again,

Erroneous pipeline warning has been solved in https://github.com/sugarlabs/locosugar/pull/14/commits/f603770d09d7ad082bd3693395e19e14112a6584.
Background image resolution is also fixed in https://github.com/sugarlabs/locosugar/pull/14/commits/bc5bc2c19d7d14cbfcf7e6410b7136054ec99b7f.
@quozl @chimosky @srevinsaju
Kindly review.

**Comment by Saumya-Mishra9129** [comment]
@JuiP can you please look here also for a release.?

**Comment by JuiP** [comment]
> @JuiP can you please look here also for a release.?

Sure! I prefer working on improvements and release after these changes are merged will save me time to rebase on this branch incase you push more commits :)

**Comment by Saumya-Mishra9129** [comment]
> Sure! I prefer working on improvements and release after these changes are merged will save me time to rebase on this branch incase you push more commits :)

Thanks! 

@quozl Please review and merge as well as I am done with changes .

**Comment by quozl** [comment]
Thanks.  I'll next be available for reviews on about 10th June, plus or minus a couple of days.

File: sprites.py [hunk 13]
```diff
@@ -119,32 +126,36 @@ def insert_in_list(self, spr, i):
...
         ''' Remove a sprite from the list. '''
         if spr in self.list:
             self.list.remove(spr)
 
-    def find_sprite(self, pos):
+    def find_sprite(self, pos, region=False):
         ''' Search based on (x, y) position. Return the 'top/first' one. '''
         list = self.list[:]
         list.reverse()
         for spr in list:
```
Comment (srevinsaju) [inline sprites.py]: I would preferably like to have any other variable than list, because `list` is a built in function of python, and that would not be good; although it does not cause any errors.

File: sprites.py [hunk 14]
```diff
@@ -182,47 +198,57 @@ def __init__(self, sprites, x, y, image):
...
-        self.rect[0] += int(pos[0])
-        self.rect[1] += int(pos[1])
+        self.rect.x += int(pos[0])
+        self.rect.y += int(pos[1])
         self.inval()
 
     def get_xy(self):
         ''' Return current (x, y) position '''
-        return (self.rect[0], self.rect[1])
+        return (self.rect.x, self.rect.y)
```
Comment (srevinsaju) [inline sprites.py]: Might be good to remove redundant parenthesis

File: sprites.py [hunk 15]
```diff
@@ -317,149 +344,143 @@ def restore(self):
...
+        # Create a new 1x1 cairo surface.
+        cs = cairo.ImageSurface(cairo.FORMAT_RGB24, 1, 1)
+        cr = cairo.Context(cs)
+        cr.set_source_surface(self.cached_surfaces[i], -x, -y)
+        cr.rectangle(0, 0, 1, 1)
+        cr.set_operator(cairo.OPERATOR_SOURCE)
+        cr.fill()
+        cs.flush()  # Ensure all the writing is done.
+        pixels = cs.get_data()  # Read the pixel.
+        return (pixels[2], pixels[1], pixels[0], 0)
```
Comment (srevinsaju) [inline sprites.py]: Missing newline at the end of file

File: LocoSugarActivity.py [hunk 16]
```diff
@@ -9,8 +9,9 @@
 # You should have received a copy of the GNU General Public License
 # along with this library; if not, write to the Free Software
 # Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA
-
-from gi.repository import Gtk, GObject, Gdk
+import gi
+gi.require_version('Gtk','3.0')
```
Comment (srevinsaju) [inline LocoSugarActivity.py:13]: Should we add
```python
gi.require_version('Gdk', '3.0')
```

File: toolbar_utils.py [hunk 17]
```diff
@@ -9,7 +9,8 @@
 # You should have received a copy of the GNU General Public License
 # along with this library; if not, write to the Free Software
 # Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA
-
+import gi
+gi.require_version('Gtk','3.0')
```
Comment (srevinsaju) [inline toolbar_utils.py]: I guess this is unnecessary as this is already called in `LocoSugarActivity.py`

**Review by srevinsaju — CHANGES_REQUESTED** [review]
Great work

I haven't tested it, some minor changes I found during a code review

File: sprites.py (hunk 14 above)
Comment (Saumya-Mishra9129) [reply sprites.py]: Yeah Thanks

File: LocoSugarActivity.py (hunk 16 above)
Comment (Saumya-Mishra9129) [reply LocoSugarActivity.py:13]: Yeah we should add.

**Comment by Saumya-Mishra9129** [comment]
> Great work
> 
> I haven't tested it, some minor changes I found during a code review

Can you test display resolution here bc5bc2c , and see if it works for you or not.

**Comment by srevinsaju** [comment]
I will hopefully check it tomorrow. Winding up for today :sleeping:

File: game.py [hunk 18]
```diff
@@ -118,10 +118,10 @@ def __init__(self, canvas, parent=None, path=None):
                     'background2.jpg']
         self._backgrounds = []
         for bg in self._BG:
+            pixbuf = GdkPixbuf.Pixbuf.new_from_file(os.path.join(self._path, 'images', bg))
+            pixbuf = pixbuf.scale_simple(self._width, self._height, GdkPixbuf.InterpType.BILINEAR)
             self._backgrounds.append(Sprite(
-                    self._sprites, 0, 0, GdkPixbuf.Pixbuf.new_from_file_at_size(
-                        os.path.join(self._path, 'images', bg),
-                        self._width, self._height)))
+                self._sprites, 0, 0, pixbuf))
```
Comment (chimosky) [inline game.py:125]: Can you explain why this change?

File: sprites.py (hunk 14 above)
Comment (chimosky) [reply sprites.py]: > Might be good to remove redundant parenthesis

`sprites.py` is maintained in `Turtle Art` activity, changing it here makes future update difficult; if you want a change it should be from there and not here.

**Comment by chimosky** [comment]
Tested, reviewed 7479cf8.

When pressing the letter that show on the loco some locos appear over other locos making it difficult to see the letter that's on that loco.

Also after typing `Hello LocoXO` I'm guessing that's where the game ends but there's no message that tells that.

Also the window for instructions extend off the screen to the right.

@Saumya-Mishra9129  Can you also apply bedb771, d77a866 upstream?

**Comment by Saumya-Mishra9129** [comment]
> @Saumya-Mishra9129 Can you also apply [bedb771](https://github.com/sugarlabs/locosugar/commit/bedb7715476273c881408f76ae8202089476cb6e), [d77a866](https://github.com/sugarlabs/locosugar/commit/d77a86624a2a8cd050f60333452b7da494fe7399) upstream?

Yeah sure. I will make changes in Turtle Art activity

> Also the window for instructions extend off the screen to the right.

I didnt get it completely how it shows to you. Can you please show me a screenshot.? Because I get it working completely fine.

> When pressing the letter that show on the loco some locos appear over other locos making it difficult to see the letter that's on that loco.

For this you can look at previous comments as I and @quozl already discussed that.You can look here as well https://github.com/sugarlabs/locosugar/issues/16.

> Also after typing Hello LocoXO I'm guessing that's where the game ends but there's no message that tells that.

Yeah there should be a feature but I guess activity is designed this way.

File: game.py (hunk 18 above)
Comment (Saumya-Mishra9129) [reply game.py:125]: The changes is to fix activity resolution, and make it compatible with all screens as it enlarges image according to screen height and width and set it.  GdkPixbuf.InterpType.BILINEAR is  interpolation mode, it has a good speed/quality balance.

**Comment by JuiP** [comment]
Hi, I'm testing on a 800*600 VM and here is a screenshot of what the display looks like.
Loco1

File: game.py (hunk 18 above)
Comment (JuiP) [reply game.py:125]: I don't think this changes anything functionally. `GdkPixbuf.Pixbuf.new_from_file_at_size()` function call immediately resizes the image after it loads from file. While `GdkPixbuf.InterpType.BILINEAR` is also used to resize image. 
Also, please tell me what is your test resolution? Did you face an issue before that making this change fixed?

**Comment by Saumya-Mishra9129** [comment]
Thanks @JuiP . The second screenshot is fine and I don't know what you are
talking about.
In 1st one , the greeting message window is slightly shifted to right. I
need to fix that only.

**Comment by JuiP** [comment]
Thanks. I actually edited my comment. Can you tell me your test display size?

**Comment by Saumya-Mishra9129** [comment]
> Thanks. I actually edited my comment. Can you tell me your test display size?

Yeah , Tested on 1920X1080 . The background image is perfect displaying in yours just the problem is with greeting message window size. We need to fix that.

File: game.py (hunk 18 above)
Comment (Saumya-Mishra9129) [reply game.py:125]: Thanks for looking. I don't understand completely btw why this change fixed the problem but as earlier It was n't working properly for large display sizes as look [here](https://github.com/sugarlabs/locosugar/pull/14#issuecomment-614414061). The problem can be according to me is that, the image we are using here is of 1200x900 pixel size, so It won't work on high resolution devices as it will automatically load 1200x900 image size, it won't scale the image. We here need to scale the image to screen size . I don't know I could be wrong. But If you can check on a device of high resolution please test there. Because I got the problem fixed there now after these changes. Inform me if I am wrong.

**Comment by JuiP** [comment]
@Saumya-Mishra9129 , I have fixed it in 88fadc8, you can cherry-pick the commit. :)

The ghost loco happened for me too. I tested master, couldnt reproduce on master.

Ghost_loco

File: game.py (hunk 18 above)
Comment (JuiP) [reply game.py:125]: line 124, mentions the size we want the image to load to. From what I understand, `GdkPixbuf.Pixbuf.new_from_file_at_size` is supposed to load the image and rescale it to the given dimensions.

**Comment by Saumya-Mishra9129** [comment]
> @Saumya-Mishra9129 , I have fixed it in [88fadc8](https://github.com/sugarlabs/locosugar/commit/88fadc833a24e8ead7e417c57fe57e4394547cc7), you can cherry-pick the commit. :)
> 
> The ghost loco happened for me too. I tested master, couldnt reproduce on master.
> 
> Ghost_loco

Thanks I need to test it. I still don't know how it is coming. However I tested 88fadc8 . It is not working for me look 
Screenshot from 2020-06-06 12-49-59

Update- look into comments of 88fadc8.

**Comment by JuiP** [comment]
Sorry, that was a reckless mistake. Fixed in 0cc356e

**Comment by Saumya-Mishra9129** [comment]
> The ghost loco happened for me too. I tested master, couldnt reproduce on master.

I am worried about it. I did subsequent rounds of testing, It happens sometimes, not every time. What I think is that at the end of a particular level, whenever a loco couldn't delete, It happens as a ghost logo. I don't think it could be because of Python3 changes. It happens in master as well.
I tested master several times, 3rd time I got a loco which couldn't get delete.

**Comment by srevinsaju** [comment]
Screenshot_20200606_120859

the ghost loco (I assume its called like that) gets overlapped over each other. Afaik, that ghost loco appears on the third or fourth level of clicking

Tested on 1024 x 768. Testing on other resolutions, will update shortly

**Comment by srevinsaju** [comment]
Before the next release, it would be good, if an opaque shape is placed behind the letters to make it readable
image

**Comment by srevinsaju** [comment]
I was not able to reproduce the ghost loco getting stuck on screen bug 



-------------

Tested on 640x480: no bugs
Screenshot_20200606_123919


-------------
Tested on 768x1366 (portrait) Looks good
Screenshot_20200606_123642

**Comment by Saumya-Mishra9129** [comment]
> I was not able to reproduce the ghost loco getting stuck on screen bug
> 
> Tested on 640x480: no bugs

> Tested on 768x1366 (portrait) Looks good


Thanks a lot @srevinsaju You did really a great work while reviewing. 

> Before the next release, it would be good, if an opaque shape is placed behind the letters to make it readable

Yeah I am thinking it as a good suggestion, we can do it, to increase visual impact on user.

**Comment by JuiP** [comment]
>  Before the next release, it would be good, if an opaque shape is placed behind the letters to make it readable
 Yeah I am thinking it as a good suggestion, we can do it, to increase visual impact on user.

I agree @Saumya-Mishra9129 @srevin. What do you think @chimosky @quozl ?

> I am worried about it. I did subsequent rounds of testing, It happens sometimes, not every time. What I think is that at the end of a particular level, whenever a loco couldn't delete, It happens as a ghost logo. I don't think it could be because of Python3 changes. It happens in master as well.
I tested master several times, 3rd time I got a loco which couldn't get delete.

Yes, you are probably right. @Saumya-Mishra9129 Have you found a way to fix it? I don't mind helping if you haven't already found a fix :)

**Comment by Saumya-Mishra9129** [comment]
> Yes, you are probably right. @Saumya-Mishra9129 Have you found a way to fix it? I don't mind helping if you haven't already found a fix :)

Thanks ,@JuiP I haven't fixed it yet. If you could do it would be great.

**Comment by srevinsaju** [comment]
@Saumya-Mishra9129 are you getting the ghost loco freqjuently; like more than 5 times while testing 10 times (approx); and is the ghost loco getting stuck at the same position?
I tried looking into it, but was unable to reproduce. I have strong suspicion if its an internal library difference like how it happened in puklanapac and FractionBounce. Are you testing it on Sugar Live Build (debian)?

**Comment by JuiP** [comment]
I have tested it 6 times,  was able to reproduce it 2 times. It might be
what you are suggesting. For reference I'm using Debian buster sugar live
build

**Comment by Saumya-Mishra9129** [comment]
> @Saumya-Mishra9129 are you getting the ghost loco freqjuently; like more than 5 times while testing 10 times (approx); and is the ghost loco getting stuck at the same position?
> I tried looking into it, but was unable to reproduce. I have strong suspicion if its an internal library difference like how it happened in puklanapac and FractionBounce. Are you testing it on Sugar Live Build (debian)?

I have tested with both packaged and live sugar . I am able to reproduce in both systems. I haven't tested puklanapac and FractionBounce.

**Comment by JuiP** [comment]
@Saumya-Mishra9129 @srevinsaju Update on the ghost loco bug - I tried reproducing it again, but couldn't. I'm not sure what has changed. I also looked at the code and couldn't find a possible reason for the ghost loco.

**Comment by Saumya-Mishra9129** [comment]
> @Saumya-Mishra9129 @srevinsaju Update on the ghost loco bug - I tried reproducing it again, but couldn't. I'm not sure what has changed. I also looked at the code and couldn't find a possible reason for the ghost loco.

Thanks for Update. Actually Why ghost loco happens and even when it happens, it's highly unpredictable here. The problem is still ambiguous for me, I also tried to look at code, I couldn't also find the reason. It also happens in master as well. So I guess better way could be to open an issue here. We can look for merge here then because the work here is completed and you can look for release also then.

**Comment by quozl** [comment]
My guess is a logic error, or API balance.  Sprites are being created, and the list of sprites is lost in some way.  It is the kind of problem that should be solved before release.  If one method of finding the cause does not work, try another method.

If you've not yet used the Python debugger, _now_ is one of those times it would be of benefit to have the skills.

**Comment by Saumya-Mishra9129** [comment]
> @Saumya-Mishra9129 @srevinsaju Update on the ghost loco bug - I tried reproducing it again, but couldn't. I'm not sure what has changed. I also looked at the code and couldn't find a possible reason for the ghost loco.

Update - @JuiP I have tried 16 times not able to reproduce ghost loco even once. I don't know what has happened now.
Last change we did is 791af63. We have also done some changes in sprites.py, maybe that has solved the problem.

**Comment by shaansubbaiah** [comment]
I found the original game trailer? at https://www.youtube .com/watch?v=0m_d8_Nfi28, the last line of the description int he site it was published says "Clicking, double-clicking, dragging and typing a few sentences makes cuckoos get better and landscapes more friendly.". 

Maybe the by landscapes more friendly they mean, less ghost locos? The links to the original windows executable and XO are dead.

EDIT: Removed the link to the site, the page is likely infected

**Comment by quozl** [comment]
Thanks.  Tested.  Approved.  No ghost loco in two cycles tested.

**Comment by JuiP** [comment]
Yes I agree, No ghost loco for me too.

**Comment by chimosky** [comment]
Reviewed, thanks.
